src/main.rs

```diff
@@ -38,31 +38,36 @@ struct MarketData {
 
 impl MarketData {
     fn insert_and_validate(&mut self, hash: &str, filerequest: &FileRequest) {
-        // let mut holders: Vec<FileRequest> = vec![];
         // check if self.files[hash] exists
         if !self.files.contains_key(hash) {
-            self.files.insert(hash.to_string(), vec![filerequest.clone()]);
+            self.files
+                .insert(hash.to_string(), vec![filerequest.clone()]);
```

Comment:
Suggestion: I think it'll be okay to actually accept the value itself rather than a reference to the value since we aren't going to use the `FileRequest` object further in where this method is called. This would avoid the unnecessary cloning and the result is still equivalent when taking ownership instead:
```Rust
fn insert_and_validate(&mut self, hash: &str, filerequest: FileRequest) {
        // check if self.files[hash] exists
        if !self.files.contains_key(hash) {
            self.files.insert(hash.to_string(), vec![]);
        }
        .
        .
        .
        producers.push(filerequest);
}
```